torch/fx/experimental/symbolic_shapes.py

```diff
@@ -500,6 +514,22 @@ def free_unbacked_symbols_with_path(
             ):
                 r[s] = path
                 pending.remove(s)
+            # When an unbacked SymInt is perfectly divisible by an integer
+            # constant, we replace it with the integer constant to improve
+            # reasoning capabilities.  However, in synthetic examples, it is
+            # then possible that the factor never is explicitly allocated.
+            # Fortunately, we can compute it by division.
+            elif (
+                isinstance(a, torch.SymInt)
+                and isinstance(s := a.node._expr, sympy.Mul)
+                and len(s.args) == 2
+                and isinstance(lhs := s.args[0], sympy.Integer)
+                and isinstance(rhs := s.args[1], sympy.Symbol)
+                and rhs in pending
```

Comment:
I think I'm a bit confused about the invariants. In my head:

(1) the main purpose of `compute_unbacked_bindings` and the related changes in this PR stack is to establish a clear source node in the FX graph for the binding site of all unbacked symints (attaching this info to `node.meta['unbacked_bindings']`)

(2) in general, we can do replacements on unbacked symints, replacing e.g. a new unbacked symint `u1` with `u0 * 4`. This means that we do **not** need to codegen a new binding for `u1`, since the replacement dictates that we can take the existing binding `u0` and blat out `u0 * 4` anywhere that we needed `u1` in the generated codegen.

I have a feeling that I'm stating (2) incorrectly: because to me, (2) implies that we do **not** need to any new unbacked bindings when we do a replacement. E.g., in an example like the below:
```
u0 = inp.item()
u1 = u0 * 4
return torch.empty(u1).view(4, -1)
```

Would not need to bother worrying about generating a binding for `u1`. (we would probably blat out `torch.empty(u0 * 4)` in the codegen)

Reply:
You are right, in your program, this code never gets executed. The code here is not for that case: it's for a more tricksy case:

```
y = x.item()
torch._check(y % 4 == 0)
... use of y ...
```

Here, we allocate u0 for y on the item() call. However, when we discover that u0 is divisible by four, we replace it with a fresh u1 = u0 * 4. This replacement is profitable because sympy is able to do more simplifications when it see s "u0 * 4" as supposed to "u1 (and oh, by the way, this is divisible by four, but not in a way sympy knows about)".

After this replacement, we now have u1 floating around everywhere. But where is u1 bound? Well oops, it's not bound anywhere: we only know it was because x.item() produces u0, and we happen to know it's divisible by 4, so we can get u1 = u0 // 4. So that's what all this logic is about.

Reply:
> we only know it was because x.item() produces u0

Hmm, I find this bit confusing. Suppose we have some code like this:
```
y: u0 = x.item()  # line 1
contrain_range(y, 16, 24)  # line 2
torch._check(y % 4 == 0)  # line 3
```
After executing line 2, our shape env should have `u0: {16, 24}`.

But once we see line 3 and do our replacement of `u1 = u0 * 4`... It would be wrong to say that `u0` corresponds to the original value from `x.item()`, no? In particular, if we had any other compile-time information about `y.item()` (like a value range), that information should now apply to the **new**` variable you created (`u1`), and not to the the `u0` in that expression (which technically would have a value range `{4, 6}` in this example, although idk if we care about actually figuring that out)

Reply:
There are some typos in your example, but a broad level, it should, and we support propagating this information to a limited extent.

First, I assume you meant x.item() everywhere you wrote y.item().

Second, your rewrite is incorrect. The rewrite is the other direction: u0 = u1 * 4. I think this explains some of the confusion on what the u0 and u1 are.

Now, in your example, prior to the rewrite, we learn 16 <= u0 <= 24. When we replace u0 = u1 * 4, we now have 16 <= u1 * 4 <= 24. From this, we can learn 4 <= u1 <= 6, thereby refining the value range on our new u1. This nets out the same, since anywhere you have u1 * 4 you have the modified range.

Reply:
Gah I confused myself. Thanks for the detailed explanation!

Reply:
Don't let Ed gaslight you, he was the one that first wrote `u1 = 4*u0`, which is not correct :P

Reply:
whooooops